proposals/1125_dynamic-supervisor-oidc-clients/README.md

```diff
@@ -0,0 +1,498 @@
+
+Perhaps the most common approach for this is to use [bcrypt](https://en.wikipedia.org/wiki/Bcrypt) with a random salt
+and a sufficiently high input cost. The salt protects against rainbow tables, and the input cost provides some
+protection against brute force guessing when the hashed password is leaked or stolen. However, the input cost also makes
+it slower for users to authenticate. The cost should be balanced against the current compute power available to
+attackers versus the inconvenience to users caused by a long pause during a genuine login attempt. There is no "best"
+value for the input cost. Even when an administrator determines a value that works for them, they should reevaluate as
+Moore's Law (and the availability of specialized hardware) catches up to their choice later.
+
+Client secrets should be decided by admins. Many OIDC Providers auto-generate client secrets and return the generated
+secret once (and only once) in their API or UI. This is good for ensuring that the secret contains a large amount of
+entropy by auto-generating long random strings using lots of possible characters. However, Pinniped has declarative
+configuration as a design goal. The configuration of Pinniped should be able to be known a priori (even before
+installing Pinniped) and should be easy to include in a Gitops workflow.
+
+Even if the client secrets are hashed with bcrypt, the hashed value is still very confidential, due to the opportunities
+for brute forcing provided by knowledge of the hashed value. Confidential data in Kubernetes should be stored in Secret
+resources. This makes it explicit that the data is confidential and many Kubernetes workflows are built on this
+assumption. For example, deployment tools will avoid showing the values in Secrets during an application deployment. As
+another example,
+[Kubernetes best practices suggest](https://kubernetes.io/docs/concepts/configuration/secret/#information-security-for-secrets)
+that admins should use authorization policies to restrict read permission to Secrets as much as possible. Additionally,
+some clusters may use the Kubernetes feature to
+[encrypt Secrets at rest](https://kubernetes.io/docs/tasks/administer-cluster/encrypt-data/), and thus reasonably expect
+that all confidential data is encrypted at rest.
+
+###### Option 1: Providing client secrets already hashed
+
+An admin could run bcrypt themselves to hash their desired client secret. Then they could write the resulting value into
+a Secret.
+
+```yaml
+apiVersion: v1
+kind: Secret
+metadata:
+  name: my-webapp-client-secret-1
+  namespace: pinniped-supervisor
+type: secrets.pinniped.dev/oidc-client-secret-bcrypt
+stringData:
+  clientSecret: $2y$10$20UQo9UzqBzT.mgDRu9TwOC...EQSbS2
+```
+
+Advantages:
+
+- Least implementation effort.
+- Admins choose their own preferred bcrypt input cost.
+- Confidential data is stored in a Secret.
+
+Disadvantages:
```

Comment:
If the aggregated API route is taken for the OIDCClient API, then client names and secrets could both be generated, correct?

Reply:
> If the aggregated API route is taken for the OIDCClient API, then client names and secrets could both be generated, correct?

Yes to both.  I modeled the secrets bit after existing Kube APIs to match their UX.  Re: names, we would probably need to split up the `metadata.name` from the generated client ID (i.e. in the same way that the secret is split up) so that you could consistently reference a particular Kube object even if you lack control over its client ID.